service/whatsapp/whatsapp.go

```diff
@@ -0,0 +1,167 @@
+		ClientToken: clientToken,
+		ServerToken: serverToken,
+		Wid:         wid,
+		EncKey:      encKey,
+		MacKey:      macKey,
+	}
+
+	session, err := s.client.RestoreWithSession(session)
+	if err != nil {
+		return fmt.Errorf("restoring session failed: %v", err)
+	}
+
+	// Save the updated session for future use without login.
+	err = writeSession(&session)
+	if err != nil {
+		return fmt.Errorf("error saving session: %v", err)
+	}
+
+	return nil
+}
+
+// LoginWithQRCode provides helper for authentication using QR code on terminal.
+// Refer: https://github.com/Rhymen/go-whatsapp#login for more information.
+func (s *Service) LoginWithQRCode() error {
+	session, err := readSession()
+	if err == nil {
+		session, err = s.client.RestoreWithSession(session)
+		if err != nil {
+			return fmt.Errorf("restoring session failed: %v", err)
+		}
+	} else {
+		// No saved session found; need to login again.
+		qr := make(chan string)
+		go func() {
+			terminal := qrcodeTerminal.New()
+			terminal.Get(<-qr).Print()
+		}()
+
+		session, err = s.client.Login(qr)
+		if err != nil {
+			return fmt.Errorf("error during login: %v", err)
+		}
+	}
+
+	err = writeSession(&session)
+	if err != nil {
+		return fmt.Errorf("error saving session: %v", err)
+	}
+
+	<-time.After(qrLoginWaitTimeSeconds * time.Second)
+
+	return nil
+}
+
+// readSession helps load saved WhatsApp session from local file system.
+func readSession() (whatsapp.Session, error) {
+	session := whatsapp.Session{}
+	file, err := os.Open(os.TempDir() + "/whatsappSession.gob")
+	if err != nil {
+		return session, err
+	}
+	defer file.Close()
+
+	decoder := gob.NewDecoder(file)
+	err = decoder.Decode(&session)
+	if err != nil {
+		return session, err
+	}
+
+	return session, nil
+}
+
+// writeSession helps save WhatsApp session to local file system.
+func writeSession(session *whatsapp.Session) error {
+	file, err := os.Create(os.TempDir() + "/whatsappSession.gob")
+	if err != nil {
+		return err
+	}
+	defer file.Close()
+
+	encoder := gob.NewEncoder(file)
+	err = encoder.Encode(*session)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// AddReceivers takes WhatsApp contacts and adds them to the internal contacts list. The Send method will send
+// a given message to all those contacts.
+func (s *Service) AddReceivers(contacts ...string) {
+	s.contacts = append(s.contacts, contacts...)
+}
+
+// Send takes a message subject and a message body and sends them to all previously set contacts.
+func (s *Service) Send(subject, message string) error {
+	if len(s.contacts) == 0 {
+		return fmt.Errorf("no contacts added as receivers")
+	}
+
+	msgText := subject + "\n" + message
+	for _, c := range s.contacts {
+		contact := c + "@s.whatsapp.net"
+		msg := whatsapp.TextMessage{
+			Info: whatsapp.MessageInfo{
+				RemoteJid: contact,
+			},
+			Text: msgText,
+		}
+
+		if _, err := s.client.Send(msg); err != nil {
+			return errors.Wrapf(err, "failed to send message to WhatsApp contact '%s'", c)
+		}
+	}
```

Comment:
```suggestion
	msg := whatsapp.TextMessage{
		Text: subject + "\n" + message,
	}

	for _, contact := range s.contacts {
		msg.Info = whatsapp.MessageInfo{
			RemoteJid: contact + "@s.whatsapp.net",
		}

		_, err := s.client.Send(msg)
		if err != nil {
			return errors.Wrapf(err, "failed to send message to WhatsApp contact '%s'", contact)
		}
	}
```

This is the only 'bigger' change. I just don't think empty contacts are a reason to error out. This might even be intended by the user for some reason. So we're gonna drop the explicit check for empty contacts in the beginning and just let the loop run. If no contacts exist we won't end up in the loop body anyways.